### PR TITLE
Validate the maximum number of route segments

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/aspnet/AspNetCore/issues/3001

I also updated the unit tests so the RoutePattern overloads are tested.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs [hunk 1]
```diff
@@ -99,6 +107,15 @@ internal static decimal ComputeOutbound(RoutePattern routePattern)
             return precedence;
         }
 
+        private static void ValidateSegementLength(int length)
+        {
+            if (length > 28)
+            {
+                // An OverflowException will be thrown by Math.Pow when greater than 28
+                throw new InvalidOperationException("Route contains too many segments and is unable to be parsed.");
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs]: We should also include the max number allowed in the error message

File: test/Microsoft.AspNetCore.Routing.Tests/Template/RoutePatternPrecedenceTests.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Routing.Patterns;
+
+namespace Microsoft.AspNetCore.Routing.Template
+{
+    public class RoutePatternPrecedenceTests : RoutePrecedenceTestsBase
```
[the remark this answers is not in the document]
Comment (JamesNK) [reply test/Microsoft.AspNetCore.Routing.Tests/Template/RoutePatternPrecedenceTests.cs]: That's all the test class does.

There are two copies of computing route precedence: one for the old template and one for pattern. This change just applies all the old template tests to pattern without copy & paste.

**Review by rynowak — APPROVED** [review]

File: test/Microsoft.AspNetCore.Routing.Tests/Template/RoutePrecedenceTestsBase.cs [hunk 3]
```diff
@@ -2,13 +2,11 @@
...
 
 using System;
-using Microsoft.Extensions.Options;
-using Moq;
 using Xunit;
 
 namespace Microsoft.AspNetCore.Routing.Template
 {
-    public class RoutePrecedenceTests
+    public abstract class RoutePrecedenceTestsBase
```
Comment (dallancarr) [inline test/Microsoft.AspNetCore.Routing.Tests/Template/RoutePrecedenceTestsBase.cs]: Tests in abstract class?

Comment (JamesNK) [reply test/Microsoft.AspNetCore.Routing.Tests/Template/RoutePrecedenceTestsBase.cs]: They are run in classes that inherit from it

**Pull request merged**